presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/JoinGraphNode.java

```diff
@@ -0,0 +1,151 @@
+
+    public JoinGraphNode(PlanNodeId id, List<PlanNode> sources, Expression filter, List<Symbol> outputSymbols)
+    {
+        super(id);
+
+        requireNonNull(sources, "sources is null");
+        requireNonNull(filter, "filters is null");
+
+        this.sources = ImmutableList.copyOf(sources);
+        this.filter = filter;
+        this.outputSymbols = ImmutableList.copyOf(outputSymbols);
+    }
+
+    public Expression getFilter()
+    {
+        return filter;
+    }
+
+    @Override
+    public <R, C> R accept(PlanVisitor<R, C> visitor, C context)
+    {
+        return visitor.visitJoinGraph(this, context);
+    }
+
+    @Override
+    public List<PlanNode> getSources()
+    {
+        return sources;
+    }
+
+    @Override
+    public List<Symbol> getOutputSymbols()
+    {
+        return outputSymbols;
+    }
+
+    @Override
+    public PlanNode replaceChildren(List<PlanNode> newChildren)
+    {
+        return new JoinGraphNode(getId(), newChildren, filter, outputSymbols);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(this.getOutputSymbols(), this.getFilter(), this.getSources().stream().map(PlanNode::getId).collect(toImmutableList()));
+    }
+
+    @Override
+    public boolean equals(Object other)
+    {
+        if (!(other instanceof JoinGraphNode)) {
+            return false;
+        }
+
+        JoinGraphNode otherJoinGraph = (JoinGraphNode) other;
+        return this.getOutputSymbols().equals(otherJoinGraph.getOutputSymbols())
+                && this.getFilter().equals(otherJoinGraph.getFilter())
+                && this.getSources().stream().map(PlanNode::getId).collect(toImmutableList()).equals(otherJoinGraph.getSources().stream().map(PlanNode::getId));
+    }
+
+    public static class JoinGraphNodeBuilder
+    {
+        private final List<PlanNode> sources = new ArrayList<>();
+        private final List<Expression> filters = new ArrayList<>();
+        private final List<Symbol> outputSymbols;
+
+        public JoinGraphNodeBuilder(JoinNode node, Lookup lookup)
+        {
+            checkState(node.getType() == INNER, "join type must be INNER");
+            this.outputSymbols = node.getOutputSymbols();
+            flattenNode(node, lookup);
+        }
+
+        private void flattenNode(PlanNode node, Lookup lookup)
+        {
+            PlanNode resolved = lookup.resolve(node);
+            if (resolved instanceof JoinNode && ((JoinNode) resolved).getType() == INNER) {
+                JoinNode joinNode = (JoinNode) resolved;
+                flattenNode(joinNode.getLeft(), lookup);
+                flattenNode(joinNode.getRight(), lookup);
+                joinNode.getCriteria().stream()
+                        .map(criterion -> new ComparisonExpression(EQUAL, criterion.getLeft().toSymbolReference(), criterion.getRight().toSymbolReference()))
+                        .forEach(filters::add);
+                joinNode.getFilter().ifPresent(filters::add);
```

Comment:
> The function gets called a different number of times depending where it is, but that would also be true if you moved any predicates above or below it.

Predicates are not freely moved around when they are nondeterministic (see `PredicatePushDown` for example). If they are moved this is considered a bug.

>  As I see it, all the rows still get filtered by random(a) no matter where you put it

The results will be significantly different. In the first case you will have either zero rows or cardinality of `c` rows. In the second case, you will have some random output cardinality between 0 and `c` size. This is semantically different. You could imagine someone doing probabilistic join and such reorder would break it.

See @pnowojski join fixup for instance: https://github.com/prestodb/presto/pull/8037